File: src/ClassifyRecord.cpp
#include <array>
#include <numeric>
#include <algorithm>
#include "../include/ClassifyRecord.h"
#include "../include/VCFFamily.h"

using namespace std;


//////////////////// ClassifyRecord ////////////////////

ClassifyRecord *ClassifyRecord::instance = nullptr;

ClassifyRecord::~ClassifyRecord() {
	for(auto p = memo.begin(); p != memo.end(); ++p) {
		delete p->second;
	}
}

ClassifyRecord *ClassifyRecord::get_instance() {
	if(!instance)
		instance = new ClassifyRecord();
	return instance;
}

pair<ParentComb, WrongType> ClassifyRecord::classify(
									const VCFFamilyRecord *record,
									const TypeDeterminer *td,
									bool one_parent) {
	const vector<int>	gts = record->unphased_gts();
	const array<int, 4>	counter = count_int_gts(gts.begin() + 2, gts.end());
	const auto	combs = td->determine(gts[0], gts[1], counter);
	return classify_record_core(combs, record->unphased_mat(),
										record->unphased_pat(), one_parent);
}

pair<ParentComb, WrongType> ClassifyRecord::classify_self_record(
												const GenoRecord *record,
												const TypeDeterminer *td) {
	const vector<int>	gts = record->unphased_gts();
	const array<int, 4>	counter = count_int_gts(gts.begin() + 1, gts.end());
	const auto	pairs = td->determine(gts[0], gts[0], counter);
	if(pairs.empty())
		return make_pair(ParentComb::PNA, WrongType::UNSPECIFIED);
	
	const int	parent_gt = gts[0];
	const ParentComb	comb = pairs[0].first;
	if(comb == ParentComb::P01x01 &&
						(parent_gt == 1 || Genotype::is_NA(parent_gt)))
		return make_pair(ParentComb::P01x01, WrongType::RIGHT);
	else if(comb == ParentComb::P00x00 &&
						(parent_gt == 0 || Genotype::is_NA(parent_gt)))
		return make_pair(ParentComb::P00x00, WrongType::RIGHT);
	else if(comb == ParentComb::P11x11 &&
						(parent_gt == 2 || Genotype::is_NA(parent_gt)))
		return make_pair(ParentComb::P11x11, WrongType::RIGHT);
	else
		return make_pair(ParentComb::PNA, WrongType::UNSPECIFIED);
}

const TypeDeterminer *ClassifyRecord::get_TypeDeterminer(size_t n) {
	auto	p = memo.find(n);
	if(p != memo.end())
		return p->second;
	
	const TypeDeterminer	*td = new TypeDeterminer(n);
	memo.insert(make_pair(n, td));
	return td;
}

// count the numbers of 0/0, 0/1, 1/1, and ./.
array<int, 4> ClassifyRecord::count_int_gts(const vector<int>& gts) const {
	return count_int_gts(gts.begin(), gts.end());
}

array<int, 4> ClassifyRecord::count_int_gts(
									vector<int>::const_iterator first,
									vector<int>::const_iterator last) const {
	array<int, 4>	ns = {0, 0, 0, 0};
	for(auto p = first; p != last; ++p) {
		ns[*p] += 1;
	}
	return ns;
}

WrongType ClassifyRecord::select_wrong_type(ParentComb comb, int mat_gt,
											int pat_gt, bool one_parent) const {
	if(TypeDeterminer::is_same_parent_genotype(comb)) {
		const int	gt = TypeDeterminer::int_gt_pair(comb).first;
		if(mat_gt == gt && pat_gt == gt)
			return WrongType::RIGHT;
		else
			return WrongType::MODIFIABLE;
	}
	else {
		// 0/0 x 0/1 -> 2, 0/0 x 1/1 -> 1, 0/1 x 1/1 -> 0
		const int	avoiding_gt = TypeDeterminer::get_avoiding_gt(comb);
		if(mat_gt == pat_gt)
			return WrongType::UNMODIFIABLE;
		else if((Genotype::is_NA(mat_gt) &&
						(!Genotype::is_NA(pat_gt) && pat_gt != avoiding_gt)) ||
				(Genotype::is_NA(pat_gt) &&
						(!Genotype::is_NA(mat_gt) && mat_gt != avoiding_gt))) {
			return one_parent ? WrongType::RIGHT : WrongType::MODIFIABLE;
		}
		else if(mat_gt != avoiding_gt && pat_gt != avoiding_gt)
			return WrongType::RIGHT;
		else
			return WrongType::MODIFIABLE;
	}
}

bool ClassifyRecord::is_matched(int mat_gt, int pat_gt, ParentComb comb) const {
	const auto	gt_pair = TypeDeterminer::int_gt_pair(comb);
	return (mat_gt == gt_pair.first && pat_gt == gt_pair.second) ||
			(pat_gt == gt_pair.first && mat_gt == gt_pair.second);
}

pair<ParentComb, WrongType> ClassifyRecord::classify_record_core(
												const vector<GTComb>& pairs,
												int mat_gt, int pat_gt,
												bool one_parent) const {
	if(pairs.size() == 1) {
		const ParentComb	p = pairs.front().first;
		return make_pair(p, select_wrong_type(p, mat_gt, pat_gt, one_parent));
	}
	else {
		return make_pair(ParentComb::PNA, WrongType::UNSPECIFIED);
	}
}

pair<ParentComb, FillType> ClassifyRecord::classify_family_record(
												const VCFFamilyRecord *record) {
	if(record->is_NA(0) || record->is_NA(1))
		return make_pair(ParentComb::PNA, FillType::IMPUTABLE);
	
	if(record->is_mat_hetero()) {
		if(record->is_pat_hetero())
			return make_pair(ParentComb::P01x01, FillType::IMPUTABLE);
		else if(record->is_00(1))
			return make_pair(ParentComb::P00x01, FillType::MAT);
		else
			return make_pair(ParentComb::P01x11, FillType::MAT);
	}
	else if(record->is_00(0)) {
		if(record->is_pat_hetero())
			return make_pair(ParentComb::P00x01, FillType::PAT);
		else if(record->is_00(1))
			return make_pair(ParentComb::P00x00, FillType::FILLED);
		else
			return make_pair(ParentComb::P00x11, FillType::FILLED);
	}
	else {
		if(record->is_pat_hetero())
			return make_pair(ParentComb::P01x11, FillType::PAT);
		else if(record->is_00(1))
			return make_pair(ParentComb::P00x11, FillType::FILLED);
		else
			return make_pair(ParentComb::P11x11, FillType::FILLED);
	}
}

File: python/ClassifyRecord.py
from __future__ import annotations

# coding: utf-8
# ClassifyRecord.py

from functools import reduce
from enum import Enum

from GenoRecord import GenoRecord
from VCFFamily import VCFFamilyRecord
from TypeDeterminer import *
from Genotype import Genotype
from common import *


#################### FillType ####################

class FillType(Enum):
	MAT = 0
	PAT = 1
	FILLED = 2
	IMPUTABLE = 3
	UNABLE = 4


#################### classify ####################

memo_tds: dict[int, TypeDeterminer] = { }

def count_int_gts(gts: list[int]) -> list[int]:
	ns = [0, 0, 0, 0]
	for gt in gts:
		ns[gt] += 1
	return ns

def classify_record(record: GenoRecord, td: TypeDeterminer,
								one_parent: bool) -> tuple[str, ParentComb]:
	gts = record.unphased_gts()
	counter = count_int_gts(gts[2:])
	pairs = td.determine(gts[0], gts[1], counter)
	pair, wrong_type = classify_record_core(pairs, gts[0], gts[1], one_parent)
	return (wrong_type, pair)

def classify_self_record(record: GenoRecord,
								td: TypeDeterminer) -> tuple[str, ParentComb]:
	gts = record.unphased_gts()
	counter = count_int_gts(gts[1:])
	pairs = td.determine(gts[0], gts[0], counter)
	if not pairs:
		return ('Unspecified', ParentComb.PNA)
	
	parent_gt = gts[0]
	comb, prob = pairs[0]
	if comb == ParentComb.P01x01 and parent_gt in (1, Genotype.NA):
		return ('Right', ParentComb.P01x01)
	elif comb == ParentComb.P00x00 and parent_gt in (0, Genotype.NA):
		return ('Right', ParentComb.P00x00)
	elif comb == ParentComb.P11x11 and parent_gt in (2, Genotype.NA):
		return ('Right', ParentComb.P11x11)
	else:
		return ('Unspecified', ParentComb.PNA)

def classify_record_core(pairs: list[tuple[ParentComb, float]],
						 mat_gt: int, pat_gt: int,
						 one_parent: bool) -> tuple[ParentComb, str]:
	def is_matched(mat_gt: int, pat_gt: int, pair: ParentComb) -> bool:
		gt_pair = pair.int_gt_pair()
		return (mat_gt, pat_gt) == gt_pair or (pat_gt, mat_gt) == gt_pair
	
	# 複数候補があれば不明
	if len(pairs) == 1:
		pair, p = pairs[0]
		if pair.is_same_parent_genotype():
			gt = pair.value >> 1
			if mat_gt == gt and pat_gt == gt:
				return (pair, 'Right')
			else:
				return (pair, 'Modifiable')
		else:
			# 0/0 x 0/1 -> 2, 0/0 x 1/1 -> 1, 0/1 x 1/1 -> 0
			avoiding_gt = (5 - pair.value) >> 1
			if mat_gt == pat_gt:
				return (pair, 'Unmodifiable')
			elif (mat_gt == Genotype.NA and
						pat_gt not in (Genotype.NA, avoiding_gt)):
				if one_parent:
					return (pair, 'Right')
				else:
					return (pair, 'Modifiable')
			elif (pat_gt == Genotype.NA and
						mat_gt not in (Genotype.NA, avoiding_gt)):
				if one_parent:
					return (pair, 'Right')
				else:
					return (pair, 'Modifiable')
			elif mat_gt != avoiding_gt and pat_gt != avoiding_gt:
				return (pair, 'Right')
			else:
				return (pair, 'Modifiable')
	else:
		return (ParentComb.PNA, 'Unspecified')

def classify_family_record(record: VCFFamilyRecord
										) -> tuple[ParentComb, FillType]:
	if record.is_mat_NA() or record.is_pat_NA():
		return (ParentComb.PNA, FillType.IMPUTABLE)
	
	i = 0 if record.is_mat_hetero() else 1
	j = 0 if record.is_pat_hetero() else 1
	if (i, j) == (0, 0):		# 0/1 x 0/1
		return (ParentComb.P01x01, FillType.IMPUTABLE)
	elif (i, j) == (1, 0):		# 0/0 x 0/1 or 1/1 x 0/1
		if record.is_00(0):
			return (ParentComb.P00x01, FillType.PAT)
		else:
			return (ParentComb.P01x11, FillType.PAT)
	elif (i, j) == (0, 1):		# 0/1 x 0/0 or 0/1 x 1/1
		if record.is_00(1):
			return (ParentComb.P00x01, FillType.MAT)
		else:
			return (ParentComb.P01x11, FillType.MAT)
	else:
		if record.is_00(0) and record.is_00(1):
			return (ParentComb.P00x00, FillType.FILLED)
		elif record.is_11(0) and record.is_11(1):
			return (ParentComb.P11x11, FillType.FILLED)
		else:													# 0/0 x 1/1
			return (ParentComb.P00x11, FillType.FILLED)

def prepare(n: int) -> None:
	if n not in memo_tds:
		td = TypeDeterminer(n)
		memo_tds[n] = td

def get_typedeterminer(n: int) -> TypeDeterminer:
	if n in memo_tds:
		return memo_tds[n]
	
	td = TypeDeterminer(n)
	memo_tds[n] = td
	return td


#################### main ####################

__all__ = ['classify_record', 'get_typedeterminer', 'prepare']
